### ISSUE TITLE
Allow Trial to use debuggers other than pdb

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5794]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)| @Julian reported|
|-|-|
|Trac ID|trac#5794|
|Type|enhancement|
|Created|2012-07-20 02:48:10Z|
|Branch|https://github.com/twisted/twisted/tree/trial-with-other-debuggers-5794|

Trial should accept an arbitrary debugger to use for debugging when the `--debug` option is passed.

Attachments:

* [debuggers.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b67/b67d122579104d8cb302434353b8f272a35d0558/84f3e13594071c76f965e94435e7eab65ee9e368.patch) (6608 bytes) - added by Julian on 2012-07-20 02:48:47Z - 
* [debuggers-take-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b67/b67d122579104d8cb302434353b8f272a35d0558/133484022b5abd27a1089ea3dc3c4110c8afffb5.patch) (9618 bytes) - added by Julian on 2012-07-20 21:25:29Z - 
* [updated_trial_debug.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b67/b67d122579104d8cb302434353b8f272a35d0558/482f0af610e0bf2da49c9169eedd9df97f3e09f1.patch) (8753 bytes) - added by cyli on 2012-09-22 21:56:23Z - 
* [debuggers-take-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b67/b67d122579104d8cb302434353b8f272a35d0558/77f6c304a17d73645c5a81ffbd0ca4e68da78276.patch) (12526 bytes) - added by Julian on 2012-09-23 02:28:01Z - 
* [debuggers-take-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b67/b67d122579104d8cb302434353b8f272a35d0558/a334c08bb271823a41a9177ee035d5ce48f452c8.patch) (12751 bytes) - added by Julian on 2012-09-23 02:31:29Z - 
* [debuggers-take-5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b67/b67d122579104d8cb302434353b8f272a35d0558/911c0371d93bf0b6dd35d63d72e5b6b68a51cfb4.patch) (12915 bytes) - added by Julian on 2012-09-23 03:11:47Z - 
* [debugger-not-found-exception.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b67/b67d122579104d8cb302434353b8f272a35d0558/e3dfe4fb864dfbcf7f1538bd91f15a45b711b792.patch) (1585 bytes) - added by Julian on 2012-10-21 00:29:28Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5794 5794
type__enhancement enhancement
reporter__Julian Julian
priority__normal normal
milestone__ 
branch__branches_trial_with_other_debuggers_5794 branches/trial-with-other-debuggers-5794
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1342752490000000 1342752490000000
changetime__1350994101000000 1350994101000000
version__None None
owner__ 
cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_2"></a>|@Julian commented|
|-|-|

There's one other hard-coded use of pdb in `twisted.python.failure` (that can also cause pdb to be run even when passing --debugger somethingelse if something fails during a Failure's __init__) that needs to be updated. I've put that in [#5795](https://github.com/twisted/twisted/issues/5795).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

Hi Julian, thanks for your patch. More options is good but your patch doesn't give any examples of which other debuggers to use, or how to use them. An addition to the howto/examples and man page would be helpful there. An arbitrary debugger is a bold statement.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_4"></a>|@Julian commented|
|-|-|

Added to man page, and a paragraph on using debuggers, which I didn't see mentioned previously at all. New patch is debuggers-take-2.patch.

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli set owner to @cyli|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli set owner to @Julian|
|-|-|

Thanks for working on this Julian!  PuDB support would be awesome!

1. Could you please add a news file describing what this patch does?  (see http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles for what one looks like and where it'd go)

2. There should be 2 new lines between class method definitions (see `test_runner.py`), as per the [twisted coding standard](http://twistedmatrix.com/trac/browser/trunk/doc/core/development/policy/coding-standard.xhtml?format=raw)

3. There is no test for the case where a debugger is not passed (the default debugger used is pdb) and the case where the debugger specified is pdb (that it gets a wrapped PDB instance)

4. It seems kind of odd to pass either a string or a debugger to the runner.  Perhaps only the string should be passed - since the runner needs to get a wrapped PDB instance anyway, reflect could also be called there.

5. It would be nice if, if pudb is not installed or the user specifies some other debugger that doesn't exist, that a friendly error message is presented to the user rather than a reflect traceback saying that there is no module named 'pudb'.

This patch is also out of date with trunk - I've tried to manually merge it in - I've included the code I used to test this (and it works!) but I may have done it wrong.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_7"></a>|@Julian removed owner|
|-|-|

Hey! Thanks for the review(s) :).

Agreement all around on all points, hope I've addressed them satisfactorily.

Didn't see *exactly* which test method my spacing was wrong on -- hope you meant that you just wanted the CapturingDebugger line also spaced by 2 lines, if not please let me know.

Thanks again (and sorry about sucking with git-svn).

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_8"></a>|@Julian commented|
|-|-|

Whoops, forgot that `diff` isn't going to include new files by default, so it left out the topfile. Sorry 'bout the noise.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli commented|
|-|-|

(In [36225]) Branching to 'trial-with-other-debuggers-5794'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_10"></a>|@cyli commented|
|-|-|

(In [36226]) Apply debuggers-take-5.patch

refs [#5794](https://github.com/twisted/twisted/issues/5794)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_11"></a>|@cyli set owner to @Julian|
|-|-|

Hi Julian!

Thanks again for working on this, and sorry the response has taken so long!  This looks pretty great - just a few comments:

1.  There is a problem with `twisted.trial.test_script.TestRun.test_debugger_not_found` when trial is run with a `--reactor` parameter.  (see [buildbot failures](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/trial-with-other-debuggers-5794)).
Since you're calling `trial.run`, it is re-parsing the arguments and trying to set the reactor twice.  You might want to patch up the config parser to not do so.

1.  Also a test that asserts that an invalid debugger raises the `_DebuggerNotFound` would be good, since that seems to the one step in that process that is missing coverage.

Thank you again for all your hard work!

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_13"></a>|@Julian removed owner|
|-|-|

Thanks for thorough reviews :).

Think this addresses the two issues.

Once again the patch is incremental.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli commented|
|-|-|

(In [36227]) Apply debugger-not-found-exception.patch

refs [#5794](https://github.com/twisted/twisted/issues/5794)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_15"></a>|@cyli commented|
|-|-|

Tests all seem to pass on [buildbot](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/trial-with-other-debuggers-5794).  Thanks again for all your work Julian!  Merging now.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_16"></a>|@cyli set status to `closed`|
|-|-|

(In [36228]) Merge trial-with-other-debuggers-5794: Allow Trial to use debuggers other than
pdb

Author: Julian
Reviewer: cyli
Fixes: [#5794](https://github.com/twisted/twisted/issues/5794)

Add a --debugger option to trial that allows the user to specify which debugger
to use if --debug is passed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

There's at least one variable in this new code that doesn't follow the naming convention: `runcall_called`.  Please make sure you're familiar with the coding standard and apply it to all code you write and review for Twisted.

Also, I would be nice to see two other follow-up improvements related to this code:

  1. What is the interface required of other debuggers used with this feature?  The prose documentation implies very vaguely that debuggers should be "similar" to pdb.  What does that mean though?  There should be an actual interface specified.
  1. Twisted has a plugin system.  It would be nice to use it here (trial already uses plugins for some of its other extensible features).
